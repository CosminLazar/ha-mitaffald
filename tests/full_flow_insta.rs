mod mqtt;

use crate::mqtt::CollectingClient;
use ha_mitaffald::{
    mitaffald::settings::{Address, AddressId, AffaldVarmeConfig},
    settings::Settings,
    sync_data,
};
use rumqttc::Publish;
use serde::{Deserialize, Serialize};
use std::time::Duration;
use testcontainers::{
    core::{ContainerPort, WaitFor},
    runners::AsyncRunner,
    GenericImage, ImageExt,
};
use url::Url;

#[tokio::test]
async fn smoke_test_insta() {
    // GenericImage::new(name, tag).with_network(network)
    let mqtt_server = GenericImage::new("hivemq/hivemq-ce", "latest")
        .with_exposed_port(ContainerPort::Tcp(1883))
        .with_wait_for(WaitFor::message_on_stdout("Started HiveMQ in"))
        .with_network("bridge")
        .start()
        .await
        .expect("Failed to start container, is Docker running?");

    let mqtt_server_port = mqtt_server
        .get_host_port_ipv4(1883)
        .await
        .expect("Failed to get port binding");

    let mut mit_affald_server = mockito::Server::new_async().await;
    let mit_affald_server_url = Url::parse(&mit_affald_server.url()).unwrap();
    let address_id = "123".to_string();
    let mit_affald_server = mit_affald_server
        .mock(
            "GET",
            format!("/api/calendar/address/{}", address_id).as_str(),
        )
        .with_status(200)
        .with_body_from_file("src/mitaffald/remote_responses/container_information.json")
        .create_async()
        .await;

    let settings = Settings {
        affaldvarme: AffaldVarmeConfig {
            address: Address::Id(AddressId { id: address_id }),
            base_url: mit_affald_server_url,
        },
        mqtt: ha_mitaffald::settings::MQTTConfig {
            client_id: "test".to_string(),
            host: "localhost".to_string(),
            port: mqtt_server_port,
            username: "".to_owned(),
            password: "".to_owned(),
        },
    };

    let mut home_assistant = CollectingClient::new();
    home_assistant.start(&settings.mqtt);

    let sync_result = sync_data(settings).await;

    assert!(
        sync_result.is_ok(),
        "Error synchronizing: {:?}",
        sync_result.err()
    );

    let ha_messages_result = home_assistant.wait_for_messages(20, Duration::from_secs(60));

    assert!(
        ha_messages_result.is_ok(),
        "Error waiting for messages: {}",
        ha_messages_result.unwrap_err()
    );

    mit_affald_server.assert_async().await;

    let actual = actual(ha_messages_result.unwrap());

    insta::with_settings!({
        filters=>vec![
            (r#"\\"last_update\\":\s*\\"\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}\.\d+\+\d{2}:\d{2}\\""#,
            r#"\"last_update\": \"[REDACTED]\""#)
        ]
    }, {
        insta::assert_yaml_snapshot!(actual);
    });
}

fn actual(messages: Vec<Publish>) -> Vec<MqttMessage> {
    let mut x: Vec<MqttMessage> = messages
        .iter()
        .map(|message| {
            let topic = message.topic.clone();
            let payload = String::from_utf8(message.payload.to_vec()).unwrap();
            MqttMessage { topic, payload }
        })
        .collect();

    x.sort();
    x
}

#[derive(Serialize, Deserialize, PartialOrd, PartialEq, Eq, Ord)]
struct MqttMessage {
    topic: String,
    payload: String,
}
