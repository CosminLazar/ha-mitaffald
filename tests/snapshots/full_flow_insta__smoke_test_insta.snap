---
source: tests/full_flow_insta.rs
expression: actual
---
- topic: garbage_bin/Glas/status
  payload: "\n            {            \n            \"name\": \"Glas\",\n            \"next_empty\": \"2024-04-18\",\n            \"last_update\": \"[REDACTED]\"\n            }"
- topic: garbage_bin/Mad__og_drikkekartoner/status
  payload: "\n            {            \n            \"name\": \"Mad- og drikkekartoner\",\n            \"next_empty\": \"2024-04-18\",\n            \"last_update\": \"[REDACTED]\"\n            }"
- topic: garbage_bin/Madaffald/status
  payload: "\n            {            \n            \"name\": \"Madaffald\",\n            \"next_empty\": \"2024-04-26\",\n            \"last_update\": \"[REDACTED]\"\n            }"
- topic: garbage_bin/Metal/status
  payload: "\n            {            \n            \"name\": \"Metal\",\n            \"next_empty\": \"2024-04-18\",\n            \"last_update\": \"[REDACTED]\"\n            }"
- topic: garbage_bin/Pap/status
  payload: "\n            {            \n            \"name\": \"Pap\",\n            \"next_empty\": \"2024-05-09\",\n            \"last_update\": \"[REDACTED]\"\n            }"
- topic: garbage_bin/Papir/status
  payload: "\n            {            \n            \"name\": \"Papir\",\n            \"next_empty\": \"2024-05-09\",\n            \"last_update\": \"[REDACTED]\"\n            }"
- topic: garbage_bin/Plast/status
  payload: "\n            {            \n            \"name\": \"Plast\",\n            \"next_empty\": \"2024-04-18\",\n            \"last_update\": \"[REDACTED]\"\n            }"
- topic: garbage_bin/Restaffald/status
  payload: "\n            {            \n            \"name\": \"Restaffald\",\n            \"next_empty\": \"2024-04-26\",\n            \"last_update\": \"[REDACTED]\"\n            }"
- topic: garbage_bin/Tekstiler/status
  payload: "\n            {            \n            \"name\": \"Tekstiler\",\n            \"next_empty\": \"2024-05-09\",\n            \"last_update\": \"[REDACTED]\"\n            }"
- topic: garbage_bin/availability
  payload: online
- topic: homeassistant/sensor/ha_affaldvarme_Glas/config
  payload: "{\n              \"object_id\": \"ha_affaldvarme_Glas\",\n              \"unique_id\": \"ha_affaldvarme_Glas\",\n              \"name\": \"Glas\",\n              \"state_topic\": \"garbage_bin/Glas/status\",\n              \"json_attributes_topic\": \"garbage_bin/Glas/status\",\n              \"value_template\": \"{{ (strptime(value_json.next_empty, '%Y-%m-%d').date() - now().date()).days }}\",\n              \"availability_topic\": \"garbage_bin/availability\",\n              \"payload_available\": \"online\",\n              \"payload_not_available\": \"offline\",\n              \"unit_of_measurement\": \"days\",\n              \"device\": {\n                \"identifiers\": [\n                  \"ha_affaldvarme\"\n                ],\n                \"name\": \"Affaldvarme integration\",\n                \"sw_version\": \"1.0\",\n                \"model\": \"Standard\",\n                \"manufacturer\": \"Your Garbage Bin Manufacturer\"\n              },\n              \"icon\": \"mdi:recycle\"\n            }"
- topic: homeassistant/sensor/ha_affaldvarme_Mad__og_drikkekartoner/config
  payload: "{\n              \"object_id\": \"ha_affaldvarme_Mad__og_drikkekartoner\",\n              \"unique_id\": \"ha_affaldvarme_Mad__og_drikkekartoner\",\n              \"name\": \"Mad- og drikkekartoner\",\n              \"state_topic\": \"garbage_bin/Mad__og_drikkekartoner/status\",\n              \"json_attributes_topic\": \"garbage_bin/Mad__og_drikkekartoner/status\",\n              \"value_template\": \"{{ (strptime(value_json.next_empty, '%Y-%m-%d').date() - now().date()).days }}\",\n              \"availability_topic\": \"garbage_bin/availability\",\n              \"payload_available\": \"online\",\n              \"payload_not_available\": \"offline\",\n              \"unit_of_measurement\": \"days\",\n              \"device\": {\n                \"identifiers\": [\n                  \"ha_affaldvarme\"\n                ],\n                \"name\": \"Affaldvarme integration\",\n                \"sw_version\": \"1.0\",\n                \"model\": \"Standard\",\n                \"manufacturer\": \"Your Garbage Bin Manufacturer\"\n              },\n              \"icon\": \"mdi:recycle\"\n            }"
- topic: homeassistant/sensor/ha_affaldvarme_Madaffald/config
  payload: "{\n              \"object_id\": \"ha_affaldvarme_Madaffald\",\n              \"unique_id\": \"ha_affaldvarme_Madaffald\",\n              \"name\": \"Madaffald\",\n              \"state_topic\": \"garbage_bin/Madaffald/status\",\n              \"json_attributes_topic\": \"garbage_bin/Madaffald/status\",\n              \"value_template\": \"{{ (strptime(value_json.next_empty, '%Y-%m-%d').date() - now().date()).days }}\",\n              \"availability_topic\": \"garbage_bin/availability\",\n              \"payload_available\": \"online\",\n              \"payload_not_available\": \"offline\",\n              \"unit_of_measurement\": \"days\",\n              \"device\": {\n                \"identifiers\": [\n                  \"ha_affaldvarme\"\n                ],\n                \"name\": \"Affaldvarme integration\",\n                \"sw_version\": \"1.0\",\n                \"model\": \"Standard\",\n                \"manufacturer\": \"Your Garbage Bin Manufacturer\"\n              },\n              \"icon\": \"mdi:recycle\"\n            }"
- topic: homeassistant/sensor/ha_affaldvarme_Metal/config
  payload: "{\n              \"object_id\": \"ha_affaldvarme_Metal\",\n              \"unique_id\": \"ha_affaldvarme_Metal\",\n              \"name\": \"Metal\",\n              \"state_topic\": \"garbage_bin/Metal/status\",\n              \"json_attributes_topic\": \"garbage_bin/Metal/status\",\n              \"value_template\": \"{{ (strptime(value_json.next_empty, '%Y-%m-%d').date() - now().date()).days }}\",\n              \"availability_topic\": \"garbage_bin/availability\",\n              \"payload_available\": \"online\",\n              \"payload_not_available\": \"offline\",\n              \"unit_of_measurement\": \"days\",\n              \"device\": {\n                \"identifiers\": [\n                  \"ha_affaldvarme\"\n                ],\n                \"name\": \"Affaldvarme integration\",\n                \"sw_version\": \"1.0\",\n                \"model\": \"Standard\",\n                \"manufacturer\": \"Your Garbage Bin Manufacturer\"\n              },\n              \"icon\": \"mdi:recycle\"\n            }"
- topic: homeassistant/sensor/ha_affaldvarme_Pap/config
  payload: "{\n              \"object_id\": \"ha_affaldvarme_Pap\",\n              \"unique_id\": \"ha_affaldvarme_Pap\",\n              \"name\": \"Pap\",\n              \"state_topic\": \"garbage_bin/Pap/status\",\n              \"json_attributes_topic\": \"garbage_bin/Pap/status\",\n              \"value_template\": \"{{ (strptime(value_json.next_empty, '%Y-%m-%d').date() - now().date()).days }}\",\n              \"availability_topic\": \"garbage_bin/availability\",\n              \"payload_available\": \"online\",\n              \"payload_not_available\": \"offline\",\n              \"unit_of_measurement\": \"days\",\n              \"device\": {\n                \"identifiers\": [\n                  \"ha_affaldvarme\"\n                ],\n                \"name\": \"Affaldvarme integration\",\n                \"sw_version\": \"1.0\",\n                \"model\": \"Standard\",\n                \"manufacturer\": \"Your Garbage Bin Manufacturer\"\n              },\n              \"icon\": \"mdi:recycle\"\n            }"
- topic: homeassistant/sensor/ha_affaldvarme_Papir/config
  payload: "{\n              \"object_id\": \"ha_affaldvarme_Papir\",\n              \"unique_id\": \"ha_affaldvarme_Papir\",\n              \"name\": \"Papir\",\n              \"state_topic\": \"garbage_bin/Papir/status\",\n              \"json_attributes_topic\": \"garbage_bin/Papir/status\",\n              \"value_template\": \"{{ (strptime(value_json.next_empty, '%Y-%m-%d').date() - now().date()).days }}\",\n              \"availability_topic\": \"garbage_bin/availability\",\n              \"payload_available\": \"online\",\n              \"payload_not_available\": \"offline\",\n              \"unit_of_measurement\": \"days\",\n              \"device\": {\n                \"identifiers\": [\n                  \"ha_affaldvarme\"\n                ],\n                \"name\": \"Affaldvarme integration\",\n                \"sw_version\": \"1.0\",\n                \"model\": \"Standard\",\n                \"manufacturer\": \"Your Garbage Bin Manufacturer\"\n              },\n              \"icon\": \"mdi:recycle\"\n            }"
- topic: homeassistant/sensor/ha_affaldvarme_Plast/config
  payload: "{\n              \"object_id\": \"ha_affaldvarme_Plast\",\n              \"unique_id\": \"ha_affaldvarme_Plast\",\n              \"name\": \"Plast\",\n              \"state_topic\": \"garbage_bin/Plast/status\",\n              \"json_attributes_topic\": \"garbage_bin/Plast/status\",\n              \"value_template\": \"{{ (strptime(value_json.next_empty, '%Y-%m-%d').date() - now().date()).days }}\",\n              \"availability_topic\": \"garbage_bin/availability\",\n              \"payload_available\": \"online\",\n              \"payload_not_available\": \"offline\",\n              \"unit_of_measurement\": \"days\",\n              \"device\": {\n                \"identifiers\": [\n                  \"ha_affaldvarme\"\n                ],\n                \"name\": \"Affaldvarme integration\",\n                \"sw_version\": \"1.0\",\n                \"model\": \"Standard\",\n                \"manufacturer\": \"Your Garbage Bin Manufacturer\"\n              },\n              \"icon\": \"mdi:recycle\"\n            }"
- topic: homeassistant/sensor/ha_affaldvarme_Restaffald/config
  payload: "{\n              \"object_id\": \"ha_affaldvarme_Restaffald\",\n              \"unique_id\": \"ha_affaldvarme_Restaffald\",\n              \"name\": \"Restaffald\",\n              \"state_topic\": \"garbage_bin/Restaffald/status\",\n              \"json_attributes_topic\": \"garbage_bin/Restaffald/status\",\n              \"value_template\": \"{{ (strptime(value_json.next_empty, '%Y-%m-%d').date() - now().date()).days }}\",\n              \"availability_topic\": \"garbage_bin/availability\",\n              \"payload_available\": \"online\",\n              \"payload_not_available\": \"offline\",\n              \"unit_of_measurement\": \"days\",\n              \"device\": {\n                \"identifiers\": [\n                  \"ha_affaldvarme\"\n                ],\n                \"name\": \"Affaldvarme integration\",\n                \"sw_version\": \"1.0\",\n                \"model\": \"Standard\",\n                \"manufacturer\": \"Your Garbage Bin Manufacturer\"\n              },\n              \"icon\": \"mdi:recycle\"\n            }"
- topic: homeassistant/sensor/ha_affaldvarme_Tekstiler/config
  payload: "{\n              \"object_id\": \"ha_affaldvarme_Tekstiler\",\n              \"unique_id\": \"ha_affaldvarme_Tekstiler\",\n              \"name\": \"Tekstiler\",\n              \"state_topic\": \"garbage_bin/Tekstiler/status\",\n              \"json_attributes_topic\": \"garbage_bin/Tekstiler/status\",\n              \"value_template\": \"{{ (strptime(value_json.next_empty, '%Y-%m-%d').date() - now().date()).days }}\",\n              \"availability_topic\": \"garbage_bin/availability\",\n              \"payload_available\": \"online\",\n              \"payload_not_available\": \"offline\",\n              \"unit_of_measurement\": \"days\",\n              \"device\": {\n                \"identifiers\": [\n                  \"ha_affaldvarme\"\n                ],\n                \"name\": \"Affaldvarme integration\",\n                \"sw_version\": \"1.0\",\n                \"model\": \"Standard\",\n                \"manufacturer\": \"Your Garbage Bin Manufacturer\"\n              },\n              \"icon\": \"mdi:recycle\"\n            }"
